aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -689,39 +689,58 @@ Tensor& _linalg_inv_out_helper_cpu(Tensor &result, Tensor& infos_lu, Tensor& inf
 // LAPACK/MAGMA/cuSOLVER error codes are saved in 'infos' tensors, they are not checked here
 static Tensor& linalg_inv_out_info(Tensor& result, Tensor& infos_lu, Tensor& infos_getri, const Tensor& input) {
   squareCheckInputs(input);
+  checkSameDevice("linalg_inv", result, input);
+  checkLinalgCompatibleDtype("linalg_inv", result, input);
+
   TORCH_INTERNAL_ASSERT(infos_lu.scalar_type() == kInt);
   TORCH_INTERNAL_ASSERT(infos_getri.scalar_type() == kInt);
-  TORCH_CHECK(result.scalar_type() == input.scalar_type(),
-    "result dtype ", result.scalar_type(), " does not match input dtype ", input.scalar_type());
-  TORCH_CHECK(result.device() == input.device(),
-    "result device ", result.device(), " does not match input device ", input.device());
+
+  bool result_input_same_type = (result.scalar_type() == input.scalar_type());
+  bool result_equal_expected_shape = result.sizes().equals(input.sizes());
+  bool is_batched_column_major = false;
+  if (result.dim() >= 2) {
+    is_batched_column_major = result.transpose(-2, -1).is_contiguous();
+  }
+
+  // if result is not empty and not in batched column major format
+  bool copy_needed = (result.numel() != 0 && !is_batched_column_major);
+  copy_needed |= !result_input_same_type;  // or result does not have the same dtype as input
+  copy_needed |= (result.numel() != 0 && !result_equal_expected_shape); // or result does not have the expected shape
+  // we have to allocate a temporary tensor
+  if (copy_needed) {
+    Tensor result_tmp = at::empty({0}, input.options());
+    result_tmp = linalg_inv_out_info(result_tmp, infos_lu, infos_getri, input);
+    at::native::resize_output(result, result_tmp.sizes());
```

Comment:
Instead of recursively calling the operation, would it be more readable to setup result_tmp either this way or as a copy of input, then check `copy_needed` before returning from this function and, if needed, perform the copy to the original result then?

Reply:
The idea is to avoid both copy and temporary tensors if possible, that's why checking `copy_needed` should happen before the actual computations and not right before the return.

I can remove the recursion.
Currently, we have a hierarchy of calls:

1. `linalg_inv` just creates out tensor and calls `linalg_inv_out`
2. `linalg_inv_out` just creates infos tensors and calls `linalg_inv_out_info`
(2) it's simple to add a variant that doesn't do `infos` synchronization and checking for error here
3. `linalg_inv_out_info` checks and prepares inputs for calling LAPACK/MAGMA/cuSOLVER
3.1 if `result` is not good for LAPACK allocate a temporary and call recursively `linalg_inv_out_info`
4. `_linalg_inv_out_helper` dispatch to LAPACK/MAGMA/cuSOLVER

So checking if the temporary tensor and copy is needed could happen in step 2, then it would need to be duplicated in the future `linalg_inv_out_without_info_check` function. I can add a separate function between 3 and 4 to avoid the recursive call.

Reply:
Sounds like it would just complicate things more to be rid of the recursive call? Probably not worth it, then.